warehouse/malware/checks/package_turnover/check.py

```diff
@@ -0,0 +1,114 @@
+
+
+class PackageTurnoverCheck(MalwareCheckBase):
+    version = 1
+    short_description = "A check for unusual changes in package ownership"
+    long_description = dedent(
+        """
+        This check looks at recently uploaded releases and determines
+        whether their owners have recently changed or decreased the security
+        of their accounts (e.g., by disabling 2FA).
+        """
+    )
+    check_type = "scheduled"
+    schedule = {"minute": 0, "hour": 0}
+
+    def __init__(self, db):
+        super().__init__(db)
+        self._scan_interval = datetime.utcnow() - timedelta(hours=24)
+
+    def user_posture_verdicts(self, project):
+        for user in project.users:
+            has_removed_2fa_method = self.db.query(
+                self.db.query(UserEvent)
+                .filter(UserEvent.user_id == user.id)
+                .filter(UserEvent.time >= self._scan_interval)
+                .filter(UserEvent.tag == "account:two_factor:method_removed")
+                .exists()
+            ).scalar()
+
+            if has_removed_2fa_method and not user.has_two_factor:
+                self.add_verdict(
+                    project_id=project.id,
+                    classification=VerdictClassification.Threat,
+                    confidence=VerdictConfidence.High,
+                    message="User with control over this package has disabled 2FA",
+                )
+
+    def user_turnover_verdicts(self, project):
+        # NOTE: This could probably be more involved to check for the case
+        # where someone adds themself, removes the real maintainers, pushes a malicious
+        # release, then reverts the ownership to the original maintainers and removes
+        # themself again.
+        recent_role_adds = (
+            self.db.query(ProjectEvent.additional)
+            .filter(ProjectEvent.project_id == project.id)
+            .filter(ProjectEvent.time >= self._scan_interval)
+            .filter(ProjectEvent.tag == "project:role:add")
+            .all()
+        )
+
+        added_users = {role_add["target_user"] for role_add, in recent_role_adds}
+        current_users = {user.username for user in project.users}
+
+        if added_users == current_users:
+            self.add_verdict(
+                project_id=project.id,
+                classification=VerdictClassification.Threat,
+                confidence=VerdictConfidence.High,
+                message="Suspicious user turnover; all current maintainers are new",
+            )
+
+    def scan(self, **kwargs):
+        prior_verdicts = (
+            self.db.query(MalwareVerdict.release_id).filter(
+                MalwareVerdict.check_id == self.id
+            )
+        ).subquery()
+
+        # NOTE: We query for releases here, but the rest of this check is
+        # project-oriented. Maybe just query for projects directly instead?
+        releases = (
+            self.db.query(Release)
```

Comment:
`self.db.query(Release.project)` perhaps?

Reply:
Hmm, I'm having some trouble getting this to work, probably because `project` is a backref and not a realized column. I could do something like this:

```
self.db.query(Release.project).join(Release.project).filter(...)
```

but this only returns a list of `bool`s. Thoughts?

Reply:
Meh, it seems fine as is. Maybe just remove/modify the comment?